File: src/js/main/commands/RichJson_merge.js
/*
    joins two structs or arrays together
*/

import {concatArrays, getFieldByKey, isJsonObject, mergeIntoTarget} from "../RichJsonHelper.js";
import {__executeRefCommand} from "./RichJson_ref.js";
import {__RICH_JSON_COMMAND_DELIMITER} from "../RichJson.js";

export function __executeMergeCommand(parser, context) {
    let refs = context.currentMember.split(__RICH_JSON_COMMAND_DELIMITER);
    let struct_or_array = getFieldByKey(parser.cache.stack, context.currentAddress);
    let currentAddress = context.currentAddress;
    context.currentMember = refs[0].trim()
    context.currentMember = __executeRefCommand(parser, context);
    context.currentAddress = currentAddress;

    if (isJsonObject(context.currentMember)) {
        mergeIntoTarget(struct_or_array, context.currentMember);
        for (let i = 1; i < refs.length; ++i) {
            context.currentMember = refs[i].trim();
            context.currentMember = __executeRefCommand(parser, context);
            context.currentAddress = currentAddress;
            mergeIntoTarget(struct_or_array, context.currentMember);
        }
    } else {
        struct_or_array = context.currentMember;
        parser.cache.stack[context.currentAddress] = struct_or_array; // TODO look ahead for array or object in __resolveRichJsonInMember
        for (let i = 1; i < refs.length; ++i) {
            context.currentMember = refs[i].trim();
            context.currentMember = __executeRefCommand(parser, context);
            context.currentAddress = currentAddress;
            struct_or_array = concatArrays(struct_or_array, context.currentMember);
            parser.cache.stack[context.currentAddress] = struct_or_array; // concat creates a new array
        }
    }

    return struct_or_array;
}

File: src/python/main/commands/rich_json_merge.py
"""
Joins two structs or arrays together
"""
from .rich_json_ref import _execute_ref_command
from ..core.rich_json import _RICH_JSON_COMMAND_DELIMITER
from ..helper.rich_json_helper import get_field, is_json_object, merge_into_target


def _execute_merge_command(parser, context):
    refs = context.current_member.split(_RICH_JSON_COMMAND_DELIMITER)
    struct_or_array = get_field(parser.cache.stack, context.current_address)
    current_address = context.current_address

    context.current_member = refs[0].strip()
    context.current_member = _execute_ref_command(parser, context)
    context.current_address = current_address

    if is_json_object(context.current_member):
        merge_into_target(struct_or_array, context.current_member)
        for i in range(1, len(refs)):
            context.current_member = refs[i].strip()
            context.current_member = _execute_ref_command(parser, context)
            context.current_address = current_address
            merge_into_target(struct_or_array, context.current_member)
    else:
        struct_or_array = []
        parser.cache.stack[context.current_address] = struct_or_array
        for i in range(0, len(refs)):
            context.current_member = refs[i].strip()
            context.current_member = _execute_ref_command(parser, context)
            context.current_address = current_address
            struct_or_array.extend(context.current_member)

    return struct_or_array
